core/modules/replica/DatabaseMySQL.cc

```diff
@@ -130,28 +121,71 @@ ConnectionParams ConnectionParams::parse(std::string const& params,
         else if ("database" == param) { connectionParams.database = val; }
         else                          {
             throw std::invalid_argument(
-                        context + "unknown parameter '" + param + "' found in the encoded parameters string: '" +
-                        params + "'");
+                    context + "unknown parameter found in the encoded parameters string");
         }
     }
     if (connectionParams.database.empty()) {
         throw std::invalid_argument(
-                context + "database name not found in the encoded parameters string: '" +
-                params + "'");
+                context + "database name not found in the encoded parameters string");
     }
     LOGS(_log, LOG_LVL_DEBUG, context << connectionParams);
 
     return connectionParams;
 }
+#else
+ConnectionParams ConnectionParams::parse(std::string const& params,
+                                         std::string const& defaultHost,
+                                         uint16_t           defaultPort,
+                                         std::string const& defaultUser,
+                                         std::string const& defaultPassword) {
+
+    static std::string const context = "ConnectionParams::parse  ";
+
+
+    std::regex re("database=([^,]*),host=([^,]*),port=([0-9]*),user=([^,]*),password=(.*)",
```

Comment:
Do you always have to specify `port=` even if using default port number? Not terribly convenient, I'd think it should be OK to skip `port=` entirely if you want to use default. Same for all other pieces. Also exact ordering of all pieces feels like a too strict requirement, why can't I write `host=localhost,database=testdb`?

Reply:
A few comments. The first one (regarding the "Great Unification of Connection Strings" :-) ). The connection string is not passed to the replication applications by itself. It's a part of a configuration method (with the corresponding parameters) specification. Presently, there are two forms of the configuration:

file:<file-path-to-config-file>
mysql:database=...

This syntax already undermines the idea of the unification.

Secondly (to the question of optional parameters), I don't believe that any production system should rely on any defaults in _this_ area (I'm not arguing against the general usefulness of defaults though).  My understanding is that in the new "bright and shiny" Kubernetes Era all parameters will be explicitly specified. I would rather push a decision on defaults up stream. This will be much safer.

Finally, I don't think putting a lot of effort into writing a complex parser for this kind of problem  would worth it. Just consider a problem of writing a regular expression which would allow the password string to appear at any part of the connection string. I'm sure this can be done, and I can spend a half-day to implement it. The current implementation is 1) reliable, and 2) it has a straightforward and strict syntax, so that a failure to provide a required parameter would result in an exception.

Reply:
@ktlim do we have such library (ready to be used from C++ not Python) available for Qserv? If that's so then I will investigate the option. Another thing is that in a design of this code I intentionally tried to avoid having a "global" parser by allowing each subclass of the Configuration service (there are two ConfigurationFile and ConfigurationMySQL) to do its own syntax checking. The seemed to be as a better option of separating concerns. At the top level (the factory method of the Configuration service) I only check for a prefix ("file:" or "mysql:") and then instantiate forward the suffix to the corresponding subclass. Introducing a top-level parser would require the base class to know about specific parameters of its subclasses. And that;s what I didn't want to do.

Though, there is an option of passing the whole configuration string to each subclass and let it take care of what's in there. I may consider this if there is an easy to use URL parser in our existing code base (LSST Stack or Qserv). Otherwise we would need to add an external package to EUPS....which may probably make a sense for other use cases.

Reply:
SQLAlchemy syntax is not that complicated, string `mysql://user:password@host:port/database` is not hard to parse even with the same sort of regex that you use (especially if you ignore IPv6 addresses and url encoding for initial implementation).
I do not understand your reasoning about defaults - your code already supports default values for everything, all I'm saying is that requiring clients to specify `database=dbname,host=,port=,user=,password=` when you only need to pass a database name looks very strange.

Reply:
I will change the regex to 'mysql://user:password@host:port/database' as per your suggestion.

Reply:
Okay, you got it! Everything is also allowed to be default. Though, I don't like a complexity of a resulting regular expression. Someone will need to understand the syntax to maintain (or change) it.
I should probably add a comprehensive unit test for this connection parser just in case if someone will modify anything in the expression or the implementation of the class. I will do this later, though. Probably in a separate ticket.
Let's close this discussion at this point. I think we've spent too much time here.